[clang-format] Allow short function body on a single line

brandb97 opened the pull request:
Fix #145161

HazardyKnusperkeks commented on clang/include/clang/Format/Format.h:
```diff
@@ -878,6 +878,12 @@ struct FormatStyle {
+  /// put on a single line. Only when AllowShortFunctionsOnASingleLine = None
+  /// and AllowShortBlocksOnASingleLine != Never, the value of this option
+  /// is true.
+  bool AllowShortFunctionBodiesOnASingleLine;
```
That way you make this an global option to be set, is this what you want to do?

If yes you need to add a `\version` and update the docs, add parsing (including tests). If no then maybe move it to the private area?

brandb97 replied:
I am not thinking of adding an global option, I will move it to the private area.

HazardyKnusperkeks commented on clang/include/clang/Format/Format.h:
```diff
@@ -878,6 +878,12 @@ struct FormatStyle {
+  /// put on a single line. Only when AllowShortFunctionsOnASingleLine = None
+  /// and AllowShortBlocksOnASingleLine != Never, the value of this option
+  /// is true.
+  bool AllowShortFunctionBodiesOnASingleLine;
```
I'm dissatisfied with the name, as am I with all the others `Allow...`, allow suggests in my opinion not, that it is forced by `clang-format` if it fits on a single line, but that it will not be touched, if it is on a single line and fits. More like the `Leave` setting we have for some options now.

brandb97 replied:
Do you think `PutShortFuncionBodiesOnASingleLine` would be a good idea?

HazardyKnusperkeks commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -532,6 +538,20 @@ class LineJoiner {
       return MergedLines;
     }
+
+    // Previously, UnwrappedLineParser would move the left brace to a new line
```
Something like `Previously` is suitable for a commit message, not for the code comment. If it needs a comment just describe what is happening. The past is irrelevant here.

brandb97 replied:
Thank you for kindly explaining how to write proper comment, I will change it in next version.

HazardyKnusperkeks commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -532,6 +538,20 @@ class LineJoiner {
+    // when AllowShortFunctionBodiesOnASingleLine is enabled. However, if the
+    // function body cannot fit on a single line, and Style.BraceWrapping.AfterFunction
+    // is false, we should merge the function name and the left brace back onto
+    // the same line
```
```suggestion
    // the same line.
```
Comments end in full stop.

brandb97 replied:
Thanks.

HazardyKnusperkeks commented on clang/lib/Format/UnwrappedLineParser.cpp:
```diff
@@ -1921,6 +1921,10 @@ void UnwrappedLineParser::parseStructuralElement(
         } else if (Style.BraceWrapping.AfterFunction) {
           addUnwrappedLine();
+        } else if (Style.AllowShortFunctionBodiesOnASingleLine) {
+          // Wrap the left brace here; we'll try to merge it back
```
Why is this needed? It seems wrong to me.

brandb97 replied:
If `Style.AllowShortFunctionBodiesOnSingleLine = true`, I modified UnwrappedLineParser to parse
```
int main() {
  //...
}
```
into unwrapped lines ["int main()", "{", ..., "}"], not ["int main() {", ..., "}"] as usual. If clang-format can't put function body on a single line, I want to merge "int main()" and "{" back if possible.

HazardyKnusperkeks reviewed (commented):
You are missing unit tests. `clang/unittests/Format/FromatTest.cpp`.

HazardyKnusperkeks commented on clang/include/clang/Format/Format.h:
```diff
@@ -3829,6 +3829,23 @@ struct FormatStyle {
+  ///   }
+  /// \endcode
+  /// \version 20
+  bool PutShortFunctionBodiesOnASingleLine;
```
In a first round you can just add `Always` and `Never` and leave `Leave` for later.

owenca commented:
Can we fix the bug instead of adding an option?

brandb97 commented:
Sorry, I promised to add enum PutShortFunctionBodiesOnASingleLine option, but I haven’t touched this code in quite a while (due to summer vacation). It seems the requirements have changed since then, so maybe it was a good thing I didn’t write it after all? ;-)

brandb97 commented:
I can just fix the bug by moving `PutShortFunctionBodiesOnASingleLine` to private area in FormatStyle and set `PutShortFunctionBodiesOnASingleLine` to `Always` if and only if `AllowShortFunctionsOnASingleLine` is `Never` and `AllowShortBlocksOnASingleLine` is `Always`. Is that what I should do?

owenca commented:
> Sorry, I promised to add enum PutShortFunctionBodiesOnASingleLine option, but I haven’t touched this code in quite a while (due to summer vacation). It seems the requirements have changed since then, so maybe it was a good thing I didn’t write it after all? ;-)

I don't think the requirements have changed from my perspective. See https://github.com/llvm/llvm-project/issues/145161#issuecomment-3125650436.


> @owenca I actually think adding a new option would be better. The reason is that when users want to specify putting function bodies on a single line, without a new option, they would have to set AllowShortFunctionsOnASingleLine: Never AllowShortBlocksOnASingleLine: Always But such a setting would also cause shorter if and while blocks to be placed on a single line, which users might not want. With a new option, this problem wouldn’t occur.

We can't keep adding all kinds of options people can come up with. See https://clang.llvm.org/docs/ClangFormatStyleOptions.html#adding-additional-style-options.

brandb97 commented:
@owenca Can you help review my code? Thank you very much.

HazardyKnusperkeks commented on clang/lib/Format/UnwrappedLineFormatter.cpp:
```diff
@@ -510,27 +515,50 @@ class LineJoiner {
       Limit -= 2;
 
       unsigned MergedLines = 0;
+      const bool NextShortBlock = NextLine.First == NextLine.Last &&
+                                  I + 3 != E && I[3]->First->is(tok::r_brace);
```
If `I + 2 == E` this is wrong.

HazardyKnusperkeks replied:
I don't understand the question, where do you get LLDB from?

brandb97 committed:
[clang-format] allow short function body on a single line

When set AllowShortBlocksOnASingleLine to Always and short function
cannot be put on a single line, clang-format doesn't tries to put
the function body on a seperate new single line. Add
tryMergeSimpleBlocks() to put short function body on a single line if
possible.

Signed-off-by: Lidong Yan <yldhome2d2@gmail.com>

brandb97 commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -14996,6 +14996,29 @@ TEST_F(FormatTest, SplitEmptyFunctionButNotRecord) {
+  verifyFormat("int foo()\n"
+               "{ return 1; }",
+               Style);
+  verifyFormat("int foo()\n"
+               "{\n"
+               "  // comment\n"
+               "  return 1;\n"
+               "}",
+               Style);
+  verifyFormat("int foo()\n"
+               "{\n"
+               "  return 1 + 1 + 1 + 1 + 1 + 1 + 1 + 1;\n"
+               "}",
+               Style);
```
Emm, I am actually making sure merge short function should be subject to column limit. But as you requested to remove, I would remove them.

owenca replied:
These test cases can pass with or without your patch, so they are probably redundant as they should already be covered in the merging simple block tests. If not, they should be added there in a separate patch.

brandb97 commented on clang/unittests/Format/FormatTest.cpp:
```diff
@@ -14996,6 +14996,29 @@ TEST_F(FormatTest, SplitEmptyFunctionButNotRecord) {
                Style);
 }
 
+TEST_F(FormatTest, SplitShortFunction) {
+  FormatStyle Style = getLLVMStyleWithColumns(40);
```
Sure, MergeShortFunctionBody seems like a better name, thanks.

owenca replied:
We've been consistently using `auto` in place of `FormatStyle` and other types for a few years now. Also, there's no need for `ColumnLimit: 40` here now.

brandb97 replied:
Got it, will fix.

brandb97 commented:
owenca ***@***.***> writes:
> 
> 
> @owenca commented on this pull request.
> In clang/unittests/Format/FormatTest.cpp:
> > +TEST_F(FormatTest, SplitShortFunction) {
> + FormatStyle Style = getLLVMStyleWithColumns(40);
> 
> We've been consistently using auto in place of FormatStyle and other types for a few years now. Also, there's no need for ColumnLimit: 40 here now.

Understand, I will just remove them.

Thanks,
Lidong
